app/services/NextUpdatesService.scala

```diff
@@ -91,30 +93,60 @@ class NextUpdatesService @Inject()(val obligationsConnector: ObligationsConnecto
   def getNextUpdates(fromDate: LocalDate, toDate: LocalDate)(implicit hc: HeaderCarrier, mtdUser: MtdItUser[_]): Future[NextUpdatesResponseModel] = {
 
     obligationsConnector.getAllObligations(fromDate, toDate).map {
-      case obligationsResponse: ObligationsModel => ObligationsModel(obligationsResponse.obligations.filter(_.obligations.nonEmpty))
+      case obligationsResponse: ObligationsModel =>
+        ObligationsModel(obligationsResponse.obligations.filter(_.obligations.nonEmpty))
       case error: NextUpdatesErrorModel => error
       case _ => NextUpdatesErrorModel(500, "Invalid response from connector")
     }
 
   }
 
-  def getQuarterlyUpdatesCounts(taxYear: TaxYear)
-                               (implicit hc: HeaderCarrier, mtdUser: MtdItUser[_]): Future[QuarterlyUpdatesCountForTaxYear]  = {
-    getNextUpdates(taxYear.toFinancialYearStart, taxYear.toFinancialYearEnd).map {
-      case obligationsModel: ObligationsModel =>
-        QuarterlyUpdatesCountForTaxYear(taxYear, obligationsModel.submissionsCount)
-      case _ => QuarterlyUpdatesCountForTaxYear(taxYear, noQuarterlyUpdates)
+  def getQuarterlyUpdatesCounts(queryTaxYear: TaxYear, optOutProposition: OptOutProposition)
+                               (implicit hc: HeaderCarrier, mtdUser: MtdItUser[_]): Future[QuarterlyUpdatesCountForTaxYear] = {
+
+    val isQueryTaxYearPreviousTaxYear = optOutProposition.availableOptOutYears.exists {
+      case y: PreviousOptOutTaxYear if y.taxYear == queryTaxYear => true
+      case _ => false
+    }
+    val isCurrentOptOutTaxYearOffered = optOutProposition.availableOptOutYears.exists {
+      case _: CurrentOptOutTaxYear => true
+      case _ => false
+    }
+
+    def getQuarterlyUpdatesCountForTaxYear(taxYearToQuery: TaxYear): Future[QuarterlyUpdatesCountForTaxYear] = {
+      getNextUpdates(taxYearToQuery.toFinancialYearStart, taxYearToQuery.toFinancialYearEnd).map {
+        case obligationsModel: ObligationsModel =>
+          QuarterlyUpdatesCountForTaxYear(queryTaxYear, obligationsModel.quarterlyUpdatesCounts)
+        case _ => QuarterlyUpdatesCountForTaxYear(queryTaxYear, noQuarterlyUpdates)
+      }
+    }
+
+    val isIncludeCountsForCurrentYear = isQueryTaxYearPreviousTaxYear && isCurrentOptOutTaxYearOffered
+
+    if (isIncludeCountsForCurrentYear) {
+
+      val sortedOffer = optOutProposition.availableTaxYearsForOptOut.sortBy(_.startYear)
+      val isDropLast = sortedOffer.size == 3
+      val yearsToProcess = if(isDropLast) sortedOffer.dropRight(LastItem) else sortedOffer
+
+      val responses = yearsToProcess.map(getQuarterlyUpdatesCountForTaxYear)
```

Comment:
This code may call the API twice for CY, when it is needed to be included as part of the count for CY-1 and the count for CY. Could this be re-written to just get the quarterly updates count for each year, and then have a separate function that processes that to generate the cumulative counts from that data?

Reply:
isn't it doing that already? i can see it already "get the quarterly updates count for each year, and then have a separate function that processes that to generate the cumulative counts from that data", happy to get on a call to discuss this may be I'm not getting the comment here, thanks

Reply:
"yearsToProcess.map(getQuarterlyUpdatesCountForTaxYear)" is only called if the queried tax year is PY and CY is offered, not "for each of those outer tax years", not for CY it won't be called and not for NY it won't be called, and it won't even be called for PY if the CY is not offered, will call to discuss

Reply:
OK, this is now resolved. :)